Updated spack 'develop' fails to build libsigesgv on RHEL6 with native gcc@4.4.7 (wrong arch)

snl-atdm-issue

I just tried upgrading Spack by pulling the Spack 'develop' version:

```
cf9de05 "multimethods: avoid calling caller_locals() in Python 3 (#13238)"
Author: Todd Gamblin <tgamblin@llnl.gov>
Date:   Thu Oct 17 06:40:23 2019 -0700 (3 hours ago)

M       lib/spack/spack/multimethod.py
M       lib/spack/spack/package.py
```

Running:

```
$ ./atdm-spack/spack/bin/spack  install  -j20
```

with this spack.yaml file:

<details>

<summary><b>spack.yaml</b></summary>

```
# Spack env file to build a compiler using Spack
#
# NOTE: Spack will not only read this file but will also write to it so make
# sure you create a copy before spack uses it!
#
spack:
  # Packages to build (ordered mostly by dependencies)
  specs: [
    gcc@7.2.0,
    ]
  modules:
    tcl:
      hash_length: 0
      naming_scheme: spack-${PACKAGE}/${VERSION}
      whitelist:
        - gcc
      blacklist:
        - '%gcc@4.4.7'
      all:
        conflict:
          - '{name}'
        filter:
          environment_blacklist: [CPATH, LIBRARY_PATH]
        environment:
          set:
            ${PACKAGE}_ROOT: ${PREFIX}
      verbose: false
  packages:
    all:
      compiler: [gcc@4.4.7]

  include:
    - packages_tools_for_compiler_build.yaml
    - common_options.yaml
```

NOTE: `packages_tools_for_compiler_build.yaml` is empty and:

`common_options.yaml` is:

```
mirrors:
  local_filesystem: file:///ascldap/users/rabartl/Spack.base3/atdm-spack/atdm-spack-tarball-mirror
config:
  misc_cache: .spack/cache
```


</details>

it showed the following error:

```
==> Installing libsigsegv
==> Searching for binary cache of libsigsegv
==> Finding buildcaches in /ascldap/users/rabartl/Spack.base3/atdm-spack/atdm-spack-tarball-mirror/build_cache
==> No binary for libsigsegv found: installing from source
curl: (37) Couldn't open file /ascldap/users/rabartl/Spack.base3/atdm-spack/atdm-spack-tarball-mirror/libsigsegv/libsigsegv-2.12.tar.gz
==> Fetching file:///ascldap/users/rabartl/Spack.base3/atdm-spack/atdm-spack-tarball-mirror/libsigsegv/libsigsegv-2.12.tar.gz
==> Fetching from file:///ascldap/users/rabartl/Spack.base3/atdm-spack/atdm-spack-tarball-mirror/libsigsegv/libsigsegv-2.12.tar.gz failed.
==> Fetching https://ftpmirror.gnu.org/libsigsegv/libsigsegv-2.12.tar.gz
==> Staging archive: /tmp/rabartl/spack-stage/spack-stage-libsigsegv-2.12-4wg47p4nrbuzpsovsxhbypwtbcu25k4t/libsigsegv-2.12.tar.gz
==> Created stage in /tmp/rabartl/spack-stage/spack-stage-libsigsegv-2.12-4wg47p4nrbuzpsovsxhbypwtbcu25k4t
==> No patches needed for libsigsegv
==> Building libsigsegv [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 77:
    '/tmp/rabartl/spack-stage/spack-stage-libsigsegv-2.12-4wg47p4nrbuzpsovsxhbypwtbcu25k4t/spack-src/configure' '--prefix=/home/rabartl/Spack.base3/atdm-spack/spack/opt/spack/linux-rhel6-haswell/gcc-4.4.7/libsigsegv-2.12-4wg47p4nrbuzpsovsxhbypwtbcu25k4t' '--enable-shared'

2 errors found in build log:
     8     checking for a thread-safe mkdir -p... /bin/mkdir -p
     9     checking for gawk... gawk
     10    checking whether make sets $(MAKE)... yes
     11    checking whether make supports nested variables... yes
     12    checking for gcc... /home/rabartl/Spack.base3/atdm-spack/spack/lib/s
           pack/env/gcc/gcc
     13    checking whether the C compiler works... no
  >> 14    configure: error: in `/tmp/rabartl/spack-stage/spack-stage-libsigseg
           v-2.12-4wg47p4nrbuzpsovsxhbypwtbcu25k4t/spack-src':
  >> 15    configure: error: C compiler cannot create executables
     16    See `config.log' for more details

See build log for details:
  /tmp/rabartl/spack-stage/spack-stage-libsigsegv-2.12-4wg47p4nrbuzpsovsxhbypwtbcu25k4t/spack-build-out.txt
```

The full output is shown below.

<details>

<summary><b>Full spack install output</b></summary>

```
==> Concretized gcc@7.2.0
 -   i446xkz  gcc@7.2.0%gcc@4.4.7~binutils languages=c,c++,fortran ~nvptx patches=51aebe82afc9a0433ca3503a36592edf5ca769b1433986919cfb9ae2ddf3e343 ~piclibs~strip arch=linux-rhel6-haswell
 -   5wln5gl      ^gmp@6.1.2%gcc@4.4.7 arch=linux-rhel6-haswell
 -   sidqv37          ^autoconf@2.69%gcc@4.4.7 arch=linux-rhel6-haswell
 -   6ebd7qb              ^m4@1.4.18%gcc@4.4.7 patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 +sigsegv arch=linux-rhel6-haswell
 -   4wg47p4                  ^libsigsegv@2.12%gcc@4.4.7 arch=linux-rhel6-haswell
 -   nqlcjvd              ^perl@5.30.0%gcc@4.4.7+cpanm+shared+threads arch=linux-rhel6-haswell
 -   ztgcm72                  ^gdbm@1.18.1%gcc@4.4.7 arch=linux-rhel6-haswell
 -   frribg6                      ^readline@8.0%gcc@4.4.7 arch=linux-rhel6-haswell
 -   cxc3lta                          ^ncurses@6.1%gcc@4.4.7~symlinks~termlib arch=linux-rhel6-haswell
 -   icmdwkr                              ^pkgconf@1.6.3%gcc@4.4.7 arch=linux-rhel6-haswell
 -   vvars5j          ^automake@1.16.1%gcc@4.4.7 arch=linux-rhel6-haswell
 -   xmpzwlz          ^libtool@2.4.6%gcc@4.4.7 arch=linux-rhel6-haswell
 -   obxrdps      ^isl@0.18%gcc@4.4.7 arch=linux-rhel6-haswell
 -   suz2poy      ^mpc@1.1.0%gcc@4.4.7 arch=linux-rhel6-haswell
 -   bkqhlr7          ^mpfr@3.1.6%gcc@4.4.7 patches=66a5d58364113a21405fc53f4a48f4e8 arch=linux-rhel6-haswell
 -   zmpvci4      ^zlib@1.2.11%gcc@4.4.7+optimize+pic+shared arch=linux-rhel6-haswell

==> Updating view at /home/rabartl/Spack.base3/.spack-env/view
==> Installing environment /home/rabartl/Spack.base3
==> Installing libsigsegv
==> Searching for binary cache of libsigsegv
==> Finding buildcaches in /ascldap/users/rabartl/Spack.base3/atdm-spack/atdm-spack-tarball-mirror/build_cache
==> No binary for libsigsegv found: installing from source
curl: (37) Couldn't open file /ascldap/users/rabartl/Spack.base3/atdm-spack/atdm-spack-tarball-mirror/libsigsegv/libsigsegv-2.12.tar.gz
==> Fetching file:///ascldap/users/rabartl/Spack.base3/atdm-spack/atdm-spack-tarball-mirror/libsigsegv/libsigsegv-2.12.tar.gz
==> Fetching from file:///ascldap/users/rabartl/Spack.base3/atdm-spack/atdm-spack-tarball-mirror/libsigsegv/libsigsegv-2.12.tar.gz failed.
==> Fetching https://ftpmirror.gnu.org/libsigsegv/libsigsegv-2.12.tar.gz
==> Staging archive: /tmp/rabartl/spack-stage/spack-stage-libsigsegv-2.12-4wg47p4nrbuzpsovsxhbypwtbcu25k4t/libsigsegv-2.12.tar.gz
==> Created stage in /tmp/rabartl/spack-stage/spack-stage-libsigsegv-2.12-4wg47p4nrbuzpsovsxhbypwtbcu25k4t
==> No patches needed for libsigsegv
==> Building libsigsegv [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 77:
    '/tmp/rabartl/spack-stage/spack-stage-libsigsegv-2.12-4wg47p4nrbuzpsovsxhbypwtbcu25k4t/spack-src/configure' '--prefix=/home/rabartl/Spack.base3/atdm-spack/spack/opt/spack/linux-rhel6-haswell/gcc-4.4.7/libsigsegv-2.12-4wg47p4nrbuzpsovsxhbypwtbcu25k4t' '--enable-shared'

2 errors found in build log:
     8     checking for a thread-safe mkdir -p... /bin/mkdir -p
     9     checking for gawk... gawk
     10    checking whether make sets $(MAKE)... yes
     11    checking whether make supports nested variables... yes
     12    checking for gcc... /home/rabartl/Spack.base3/atdm-spack/spack/lib/s
           pack/env/gcc/gcc
     13    checking whether the C compiler works... no
  >> 14    configure: error: in `/tmp/rabartl/spack-stage/spack-stage-libsigseg
           v-2.12-4wg47p4nrbuzpsovsxhbypwtbcu25k4t/spack-src':
  >> 15    configure: error: C compiler cannot create executables
     16    See `config.log' for more details

See build log for details:
  /tmp/rabartl/spack-stage/spack-stage-libsigsegv-2.12-4wg47p4nrbuzpsovsxhbypwtbcu25k4t/spack-build-out.txt
```

</details>

--- a/lib/spack/llnl/util/cpu/microarchitectures.json
+++ b/lib/spack/llnl/util/cpu/microarchitectures.json
@@ -56,7 +56,7 @@
             "flags": "-march={name} -mtune=generic"
           },
           {
-            "versions": "4.0:4.1.2",
+            "versions": ":4.1.2",
             "name": "x86-64",
             "flags": "-march={name} -mtune={name}"
           }
@@ -79,7 +79,7 @@
       ],
       "compilers": {
         "gcc": {
-          "versions": "4:",
+          "versions": "4.0.4:",
           "flags": "-march={name} -mtune={name}"
         },
         "clang": {
@@ -100,7 +100,7 @@
       ],
       "compilers": {
         "gcc": {
-          "versions": "4:",
+          "versions": "4.3.0:",
           "flags": "-march={name} -mtune={name}"
         },
         "clang": {
@@ -225,7 +225,7 @@
             "flags": "-march={name} -mtune={name}"
           },
           {
-            "versions": ":4.8.5",
+            "versions": "4.6:4.8.5",
             "name": "core-avx-i",
             "flags": "-march={name} -mtune={name}"
           }
@@ -266,7 +266,7 @@
             "flags": "-march={name} -mtune={name}"
           },
           {
-            "versions": ":4.8.5",
+            "versions": "4.8:4.8.5",
             "name": "core-avx2",
             "flags": "-march={name} -mtune={name}"
           }
@@ -343,7 +343,7 @@
       ],
       "compilers": {
         "gcc": {
-          "versions": "5.3:",
+          "versions": "6.0:",
           "flags": "-march={name} -mtune={name}"
         },
         "clang": {
@@ -432,7 +432,7 @@
       "compilers": {
         "gcc": {
           "name": "skylake-avx512",
-          "versions": "5.3:",
+          "versions": "6.0:",
           "flags": "-march={name} -mtune={name}"
         },
         "clang": {
@@ -481,7 +481,7 @@
       ],
       "compilers": {
         "gcc": {
-          "versions": "8:",
+          "versions": "8.0:",
           "flags": "-march={name} -mtune={name}"
         },
         "clang": {
@@ -527,7 +527,7 @@
       ],
       "compilers": {
         "gcc": {
-          "versions": "9:",
+          "versions": "9.0:",
           "flags": "-march={name} -mtune={name}"
         },
         "clang": {
@@ -588,7 +588,7 @@
       "compilers": {
         "gcc": {
           "name": "icelake-client",
-          "versions": "8:",
+          "versions": "8.0:",
           "flags": "-march={name} -mtune={name}"
         },
         "clang": [
@@ -628,7 +628,7 @@
       "compilers": {
         "gcc": {
           "name": "bdver1",
-          "versions": "4.6:",
+          "versions": "4.7:",
           "flags": "-march={name} -mtune={name}"
         },
         "clang": {
@@ -788,7 +788,7 @@
       "compilers": {
         "gcc": {
           "name": "znver1",
-          "versions": "6:",
+          "versions": "6.0:",
           "flags": "-march={name} -mtune={name}"
         },
         "clang": {
@@ -833,7 +833,7 @@
       "compilers": {
         "gcc": {
           "name": "znver2",
-          "versions": "9:",
+          "versions": "9.0:",
           "flags": "-march={name} -mtune={name}"
         },
         "clang": {
@@ -851,7 +851,7 @@
       "compilers": {
         "gcc": {
           "name": "powerpc64",
-          "versions": "4:",
+          "versions": ":",
           "flags": "-mcpu={name} -mtune={name}"
         },
         "clang": {
@@ -868,7 +868,7 @@
       "features": [],
       "compilers": {
         "gcc": {
-          "versions": "4.5:",
+          "versions": "4.4:",
           "flags": "-mcpu={name} -mtune={name}"
         },
         "clang": {
@@ -911,7 +911,7 @@
       "features": [],
       "compilers": {
         "gcc": {
-          "versions": "6:",
+          "versions": "6.0:",
           "flags": "-mcpu={name} -mtune={name}"
         },
         "clang": {
@@ -929,7 +929,7 @@
       "compilers": {
         "gcc": {
           "name": "powerpc64le",
-          "versions": "4:",
+          "versions": "4.8:",
           "flags": "-mcpu={name} -mtune={name}"
         },
         "clang": {
@@ -974,7 +974,7 @@
       "compilers": {
         "gcc": {
           "name": "power9",
-          "versions": "6:",
+          "versions": "6.0:",
           "flags": "-mcpu={name} -mtune={name}"
         },
         "clang": {
@@ -991,7 +991,7 @@
       "features": [],
       "compilers": {
         "gcc": {
-          "versions": "4:",
+          "versions": "4.8.0:",
           "flags": "-march=armv8-a -mtune=generic"
         },
         "clang": {

--- a/lib/spack/spack/test/architecture.py
+++ b/lib/spack/spack/test/architecture.py
@@ -173,7 +173,7 @@ def test_arch_spec_container_semantic(item, architecture_str):
 
 @pytest.mark.parametrize('compiler_spec,target_name,expected_flags', [
     # Check compilers with version numbers from a single toolchain
-    ('gcc@4.7.2', 'haswell', '-march=core-avx2 -mtune=core-avx2'),
+    ('gcc@4.7.2', 'ivybridge', '-march=core-avx-i -mtune=core-avx-i'),
     # Check mixed toolchains
     ('clang@8.0.0', 'broadwell', ''),
     # Check clang compilers with 'apple' suffix

--- a/lib/spack/spack/test/concretize.py
+++ b/lib/spack/spack/test/concretize.py
@@ -584,11 +584,13 @@ class TestConcretize(object):
             Spec(spec).concretized()
 
     @pytest.mark.parametrize('spec, best_achievable', [
+        ('mpileaks%gcc@4.4.7', 'core2'),
         ('mpileaks%gcc@4.8', 'haswell'),
-        ('mpileaks%gcc@5.3.0', 'skylake_avx512'),
+        ('mpileaks%gcc@5.3.0', 'broadwell'),
         # Apple's clang always falls back to x86-64 for now
         ('mpileaks%clang@9.1.0-apple', 'x86_64')
     ])
+    @pytest.mark.regression('13361')
     def test_adjusting_default_target_based_on_compiler(
             self, spec, best_achievable, current_host
     ):

--- a/lib/spack/spack/test/modules/lmod.py
+++ b/lib/spack/spack/test/modules/lmod.py
@@ -264,7 +264,7 @@ class TestLmod(object):
         assert writer.conf.core_compilers
 
     @pytest.mark.parametrize('spec_str', [
-        'mpileaks target=haswell',
+        'mpileaks target=nocona',
         'mpileaks target=core2',
         'mpileaks target=x86_64',
     ])

